File: net/src/Sdl.Web.PublicContentApi/PublicContentApi.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Sdl.Web.PublicContentApi.ContentModel;
using System.Threading;
using Microsoft.CSharp.RuntimeBinder;
using Sdl.Web.GraphQLClient;
using Sdl.Web.GraphQLClient.Request;
using Sdl.Web.GraphQLClient.Response;
using Sdl.Web.GraphQLClient.Schema;
using Sdl.Web.HttpClient;
using Sdl.Web.PublicContentApi.Utils;
using Sdl.Web.Core;
using Sdl.Web.PublicContentApi.Exceptions;

namespace Sdl.Web.PublicContentApi
{
    /// <summary>
    /// Public Content Api
    /// </summary>
    public class PublicContentApi : IGraphQLClient, IPublicContentApi, IPublicContentApiAsync
    {
        private readonly IGraphQLClient _client;
        public ILogger Logger { get; } = new NullLogger();

        public PublicContentApi(IGraphQLClient graphQLclient)
        {
            _client = graphQLclient;
        }

        public PublicContentApi(IGraphQLClient graphQLclient, ILogger logger)
        {
            _client = graphQLclient;
            Logger = logger ?? new NullLogger();
        }

        public GraphQLSchema Schema => _client.Schema;

        public async Task<GraphQLSchema> SchemaAsync() => await _client.SchemaAsync();

        #region IGraphQLClient

        public int Timeout
        {
            get { return _client.Timeout; }
            set { _client.Timeout = value; }
        }

        public IHttpClient HttpClient
            => _client.HttpClient;

        public IGraphQLResponse Execute(IGraphQLRequest request)
            => _client.Execute(request);

        public IGraphQLTypedResponse<T> Execute<T>(IGraphQLRequest request)
            => _client.Execute<T>(request);

        public async Task<IGraphQLResponse> ExecuteAsync(IGraphQLRequest request, CancellationToken cancellationToken)
            => await _client.ExecuteAsync(request, cancellationToken);

        public async Task<IGraphQLTypedResponse<T>> ExecuteAsync<T>(IGraphQLRequest request,
            CancellationToken cancellationToken)
            => await _client.ExecuteAsync<T>(request, cancellationToken);

        #endregion

        #region IPublicContentApi

        public IContextData GlobalContextData { get; set; } = new ContextData();

        public Page GetPage(ContentNamespace ns, int publicationId, int pageId, string customMetaFilter, IContextData contextData) 
            => _client.Execute<ContentQuery>(
                GraphQLRequests.Page(ns, publicationId, pageId, customMetaFilter, contextData, GlobalContextData))
                .TypedResponseData.Page;

        public Page GetPage(ContentNamespace ns, int publicationId, string url, string customMetaFilter,
            IContextData contextData)
            => _client.Execute<ContentQuery>(
                GraphQLRequests.Page(ns, publicationId, url, customMetaFilter, contextData, GlobalContextData))
                .TypedResponseData.Page;

        public Page GetPage(ContentNamespace ns, int publicationId, CmUri cmUri, string customMetaFilter,
            IContextData contextData)
            => _client.Execute<ContentQuery>(
                GraphQLRequests.Page(ns, publicationId, cmUri, customMetaFilter, contextData, GlobalContextData))
                .TypedResponseData.Page;

        public BinaryComponent GetBinaryComponent(ContentNamespace ns, int publicationId, int binaryId,
            IContextData contextData) => _client.Execute<ContentQuery>(
                GraphQLRequests.BinaryComponent(ns, publicationId, binaryId, contextData, GlobalContextData))
                .TypedResponseData.BinaryComponent;

        public BinaryComponent GetBinaryComponent(ContentNamespace ns, int publicationId, string url,
            IContextData contextData)
            => _client.Execute<ContentQuery>(GraphQLRequests.BinaryComponent(ns, publicationId, url, contextData,
                GlobalContextData)).TypedResponseData.BinaryComponent;

        public BinaryComponent GetBinaryComponent(CmUri cmUri, IContextData contextData)
            => _client.Execute<ContentQuery>(GraphQLRequests.BinaryComponent(cmUri, contextData, GlobalContextData))
                .TypedResponseData.BinaryComponent;

        public ItemConnection ExecuteItemQuery(InputItemFilter filter, InputSortParam sort, IPagination pagination,
            string customMetaFilter, bool renderContent, bool includeContainerItems, IContextData contextData)
            =>
                _client.Execute<ContentQuery>(GraphQLRequests.ExecuteItemQuery(filter, sort, pagination,
                    customMetaFilter, renderContent, includeContainerItems,
                    contextData, GlobalContextData)).TypedResponseData.Items;

        public Publication GetPublication(ContentNamespace ns, int publicationId,
            string customMetaFilter, IContextData contextData)
            => _client.Execute<ContentQuery>(GraphQLRequests.Publication(ns, publicationId, customMetaFilter,
                contextData, GlobalContextData)).TypedResponseData.Publication;

        public PublicationConnection GetPublications(ContentNamespace ns, IPagination pagination,
            InputPublicationFilter filter, string customMetaFilter,
            IContextData contextData)
            =>
                _client.Execute<ContentQuery>(GraphQLRequests.Publications(ns, pagination, filter, customMetaFilter,
                    contextData,
                    GlobalContextData)).TypedResponseData.Publications;

        public string ResolvePageLink(ContentNamespace ns, int publicationId, int pageId, bool renderRelativeLink = true)
            =>
                _client.Execute<ContentQuery>(GraphQLRequests.ResolvePageLink(ns, publicationId, pageId,
                    renderRelativeLink))
                    .TypedResponseData.PageLink.Url;

        public string ResolveComponentLink(ContentNamespace ns, int publicationId, int componentId, int? sourcePageId,
            int? excludeComponentTemplateId, bool renderRelativeLink = true)
            => _client.Execute<ContentQuery>(GraphQLRequests.ResolveComponentLink(ns, publicationId, componentId,
                sourcePageId, excludeComponentTemplateId, renderRelativeLink)).TypedResponseData.ComponentLink.Url;

        public string ResolveBinaryLink(ContentNamespace ns, int publicationId, int binaryId, string variantId,
            bool renderRelativeLink = true)
            =>
                _client.Execute<ContentQuery>(GraphQLRequests.ResolveBinaryLink(ns, publicationId, binaryId, variantId,
                    renderRelativeLink))
                    .TypedResponseData.BinaryLink.Url;

        public string ResolveDynamicComponentLink(ContentNamespace ns, int publicationId, int pageId, int componentId,
            int templateId, bool renderRelativeLink = true)
            => _client.Execute<ContentQuery>(GraphQLRequests.ResolveDynamicComponentLink(ns, publicationId, pageId,
                componentId, templateId, renderRelativeLink)).TypedResponseData.DynamicComponentLink.Url;

        public PublicationMapping GetPublicationMapping(ContentNamespace ns, string url)
            => _client.Execute<ContentQuery>(GraphQLRequests.PublicationMapping(ns, url))
                .TypedResponseData.PublicationMapping;

        public dynamic GetPageModelData(ContentNamespace ns, int publicationId, int pageId, ContentType contentType,
            DataModelType modelType, PageInclusion pageInclusion, bool renderContent, IContextData contextData)
        {
            try
            {
                var response =
                    _client.Execute(GraphQLRequests.PageModelData(ns, publicationId, pageId, contentType, modelType,
                        pageInclusion, renderContent, contextData, GlobalContextData));
                return response.Data.page.rawContent.data;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get page model data (namespaceId:{ns}, publicationId:{publicationId}, pageId:{pageId}",
                    e);
            }
        }

        public dynamic GetPageModelData(ContentNamespace ns, int publicationId, string url, ContentType contentType,
            DataModelType modelType, PageInclusion pageInclusion, bool renderContent, IContextData contextData)
        {
            try
            {
                var response =
                    _client.Execute(GraphQLRequests.PageModelData(ns, publicationId, url, contentType, modelType,
                        pageInclusion, renderContent, contextData, GlobalContextData));
                return response.Data.page.rawContent.data;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get page model data (namespaceId:{ns}, publicationId:{publicationId}, url:{url}", e);
            }
        }

        public dynamic GetEntityModelData(ContentNamespace ns, int publicationId, int entityId, int templateId,
            ContentType contentType,
            DataModelType modelType, DcpType dcpType, bool renderContent, IContextData contextData)
        {
            try
            {
                var response =
                    _client.Execute(GraphQLRequests.EntityModelData(ns, publicationId, entityId, templateId, contentType,
                        modelType,
                        dcpType, renderContent, contextData, GlobalContextData));
                return response.Data.componentPresentation.rawContent.data;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get enity model data (namespaceId:{ns}, publicationId:{publicationId}, entityId:{entityId}",
                    e);
            }
        }

        public TaxonomySitemapItem GetSitemap(ContentNamespace ns, int publicationId, int descendantLevels,
            IContextData contextData)
        {
            try
            {
                var response =
                    _client.Execute<ContentQuery>(GraphQLRequests.Sitemap(ns, publicationId, descendantLevels,
                        contextData, GlobalContextData));
                return response.TypedResponseData.Sitemap;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get sitemap (namespaceId:{ns}, publicationId:{publicationId}, descendantLevels:{descendantLevels}",
                    e);
            }
        }

        public List<TaxonomySitemapItem> GetSitemapSubtree(ContentNamespace ns, int publicationId, string taxonomyNodeId,
            int descendantLevels, bool includeAncestors,
            IContextData contextData)
        {
            try
            {
                var response =
                    _client.Execute<ContentQuery>(GraphQLRequests.SitemapSubtree(ns, publicationId, taxonomyNodeId,
                        descendantLevels, includeAncestors, contextData, GlobalContextData));
                return response.TypedResponseData.SitemapSubtree;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get sitemap subtree (namespaceId:{ns}, publicationId:{publicationId}, taxonomyNodeId{taxonomyNodeId}, descendantLevels:{descendantLevels}",
                    e);
            }
        }

        #endregion

        #region IPublicContentApiAsync

        public async Task<Page> GetPage(ContentNamespace ns, int publicationId, int pageId,
            string customMetaFilter, IContextData contextData,
            CancellationToken cancellationToken = default(CancellationToken))
            => (await _client.ExecuteAsync<ContentQuery>(
                GraphQLRequests.Page(ns, publicationId, pageId, customMetaFilter, contextData, GlobalContextData),
                cancellationToken))
                .TypedResponseData.Page;

        public async Task<Page> GetPage(ContentNamespace ns, int publicationId, string url, string customMetaFilter,
            IContextData contextData, CancellationToken cancellationToken = default(CancellationToken))
            => (await _client.ExecuteAsync<ContentQuery>(
                GraphQLRequests.Page(ns, publicationId, url, customMetaFilter, contextData, GlobalContextData),
                cancellationToken))
                .TypedResponseData.Page;

        public async Task<Page> GetPage(ContentNamespace ns, int publicationId, CmUri cmUri,
            string customMetaFilter, IContextData contextData,
            CancellationToken cancellationToken = default(CancellationToken))
            => (await _client.ExecuteAsync<ContentQuery>(
                GraphQLRequests.Page(ns, publicationId, cmUri, customMetaFilter, contextData, GlobalContextData),
                cancellationToken))
                .TypedResponseData.Page;

        public async Task<BinaryComponent> GetBinaryComponentAsync(ContentNamespace ns, int publicationId, int binaryId,
            IContextData contextData, CancellationToken cancellationToken = default(CancellationToken)) => (await
                _client.ExecuteAsync<ContentQuery>(
                    GraphQLRequests.BinaryComponent(ns, publicationId, binaryId, contextData, GlobalContextData),
                    cancellationToken)).TypedResponseData.BinaryComponent;

        public async Task<BinaryComponent> GetBinaryComponentAsync(ContentNamespace ns, int publicationId, string url,
            IContextData contextData, CancellationToken cancellationToken = default(CancellationToken)) => (await
                _client.ExecuteAsync<ContentQuery>(
                    GraphQLRequests.BinaryComponent(ns, publicationId, url, contextData, GlobalContextData),
                    cancellationToken)).TypedResponseData.BinaryComponent;

        public async Task<BinaryComponent> GetBinaryComponentAsync(CmUri cmUri,
            IContextData contextData, CancellationToken cancellationToken = default(CancellationToken)) => (await
                _client.ExecuteAsync<ContentQuery>(
                    GraphQLRequests.BinaryComponent(cmUri, contextData, GlobalContextData), cancellationToken))
                .TypedResponseData.BinaryComponent;

        public async Task<ItemConnection> ExecuteItemQueryAsync(InputItemFilter filter, InputSortParam sort,
            IPagination pagination, string customMetaFilter, bool renderContent, bool includeContainerItems,
            IContextData contextData,
            CancellationToken cancellationToken = default(CancellationToken)) => (
                await
                    _client.ExecuteAsync<ContentQuery>(
                        GraphQLRequests.ExecuteItemQuery(filter, sort, pagination, customMetaFilter, renderContent,
                            includeContainerItems,
                            contextData, GlobalContextData)
                        , cancellationToken)).TypedResponseData.Items;

        public async Task<Publication> GetPublicationAsync(ContentNamespace ns, int publicationId,
            string customMetaFilter,
            IContextData contextData, CancellationToken cancellationToken = default(CancellationToken)) => (
                await
                    _client.ExecuteAsync<ContentQuery>(
                        GraphQLRequests.Publication(ns, publicationId, customMetaFilter, contextData, GlobalContextData),
                        cancellationToken)).TypedResponseData.Publication;

        public async Task<PublicationConnection> GetPublicationsAsync(ContentNamespace ns, IPagination pagination,
            InputPublicationFilter filter, string customMetaFilter, IContextData contextData,
            CancellationToken cancellationToken = default(CancellationToken))
            =>
                (await
                    _client.ExecuteAsync<ContentQuery>(
                        GraphQLRequests.Publications(ns, pagination, filter, customMetaFilter, contextData,
                            GlobalContextData), cancellationToken)).TypedResponseData.Publications;


        public async Task<string> ResolvePageLinkAsync(ContentNamespace ns, int publicationId, int pageId,
            bool renderRelativeLink = true,
            CancellationToken cancellationToken = default(CancellationToken)) => (
                await
                    _client.ExecuteAsync<ContentQuery>(
                        GraphQLRequests.ResolvePageLink(ns, publicationId, pageId, renderRelativeLink),
                        cancellationToken)).TypedResponseData.PageLink.Url;

        public async Task<string> ResolveComponentLinkAsync(ContentNamespace ns, int publicationId, int componentId,
            int? sourcePageId,
            int? excludeComponentTemplateId, bool renderRelativeLink = true,
            CancellationToken cancellationToken = default(CancellationToken)) => (
                await
                    _client.ExecuteAsync<ContentQuery>(
                        GraphQLRequests.ResolveComponentLink(ns, publicationId, componentId, sourcePageId,
                            excludeComponentTemplateId, renderRelativeLink), cancellationToken)).TypedResponseData
                .ComponentLink.Url;

        public async Task<string> ResolveBinaryLinkAsync(ContentNamespace ns, int publicationId, int binaryId,
            string variantId, bool renderRelativeLink = true,
            CancellationToken cancellationToken = default(CancellationToken)) => (
                await
                    _client.ExecuteAsync<ContentQuery>(
                        GraphQLRequests.ResolveBinaryLink(ns, publicationId, binaryId, variantId, renderRelativeLink),
                        cancellationToken))
                .TypedResponseData.BinaryLink.Url;

        public async Task<string> ResolveDynamicComponentLinkAsync(ContentNamespace ns, int publicationId, int pageId,
            int componentId,
            int templateId, bool renderRelativeLink = true,
            CancellationToken cancellationToken = default(CancellationToken)) => (
                await
                    _client.ExecuteAsync<ContentQuery>(
                        GraphQLRequests.ResolveDynamicComponentLink(ns, publicationId, pageId, componentId, templateId,
                            renderRelativeLink),
                        cancellationToken)).TypedResponseData.DynamicComponentLink.Url;

        public async Task<PublicationMapping> GetPublicationMappingAsync(ContentNamespace ns, string url,
            CancellationToken cancellationToken = default(CancellationToken)) => (await
                _client.ExecuteAsync<ContentQuery>(GraphQLRequests.PublicationMapping(ns, url), cancellationToken))
                .TypedResponseData.PublicationMapping;

        public async Task<dynamic> GetPageModelDataAsync(ContentNamespace ns, int publicationId, int pageId,
            ContentType contentType,
            DataModelType modelType, PageInclusion pageInclusion, bool renderContent, IContextData contextData,
            CancellationToken cancellationToken = default(CancellationToken))
        {
            try
            {
                var response =
                    await
                        _client.ExecuteAsync(
                            GraphQLRequests.PageModelData(ns, publicationId, pageId, contentType, modelType,
                                pageInclusion, renderContent, contextData, GlobalContextData), cancellationToken);
                return response.Data.page.rawContent.data;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get page model data (namespaceId:{ns}, publicationId:{publicationId}, pageId:{pageId}",
                    e);
            }
        }

        public async Task<dynamic> GetPageModelDataAsync(ContentNamespace ns, int publicationId, string url,
            ContentType contentType,
            DataModelType modelType, PageInclusion pageInclusion, bool renderContent, IContextData contextData,
            CancellationToken cancellationToken = default(CancellationToken))
        {
            try
            {
                var response =
                    await
                        _client.ExecuteAsync(
                            GraphQLRequests.PageModelData(ns, publicationId, url, contentType, modelType, pageInclusion,
                                renderContent, contextData, GlobalContextData), cancellationToken);
                return response.Data.page.rawContent.data;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get page model data (namespaceId:{ns}, publicationId:{publicationId}, url:{url}", e);
            }
        }

        public async Task<dynamic> GetEntityModelDataAsync(ContentNamespace ns, int publicationId, int entityId,
            int templateId,
            ContentType contentType,
            DataModelType modelType, DcpType dcpType, bool renderContent, IContextData contextData,
            CancellationToken cancellationToken = default(CancellationToken))
        {
            try
            {
                var response =
                    await
                        _client.ExecuteAsync(
                            GraphQLRequests.EntityModelData(ns, publicationId, entityId, templateId, contentType,
                                modelType, dcpType,
                                renderContent, contextData, GlobalContextData), cancellationToken);
                return response.Data.entity.rawContent.data;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get enity model data (namespaceId:{ns}, publicationId:{publicationId}, entityId:{entityId}",
                    e);
            }
        }

        public async Task<TaxonomySitemapItem> GetSitemapAsync(ContentNamespace ns, int publicationId,
            int descendantLevels, IContextData contextData,
            CancellationToken cancellationToken = default(CancellationToken))
        {
            try
            {
                var response =
                    await
                        _client.ExecuteAsync(
                            GraphQLRequests.Sitemap(ns, publicationId, descendantLevels, contextData, GlobalContextData),
                            cancellationToken);
                return response.Data.sitemap;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get sitemap (namespaceId:{ns}, publicationId:{publicationId}, descendantLevels:{descendantLevels}",
                    e);
            }
        }

        public async Task<List<TaxonomySitemapItem>> GetSitemapSubtreeAsync(ContentNamespace ns, int publicationId,
            string taxonomyNodeId, int descendantLevels, bool includeAncestors,
            IContextData contextData, CancellationToken cancellationToken = default(CancellationToken))
        {
            try
            {
                var response =
                    await
                        _client.ExecuteAsync(
                            GraphQLRequests.SitemapSubtree(ns, publicationId, taxonomyNodeId, descendantLevels,
                                includeAncestors, contextData, GlobalContextData), cancellationToken);
                return response.Data.sitemapSubtree;
            }
            catch (RuntimeBinderException e)
            {
                throw new PcaException(
                    $"Failed to get sitemap subtree (namespaceId:{ns}, publicationId:{publicationId}, taxonomyNodeId{taxonomyNodeId}, descendantLevels:{descendantLevels}",
                    e);
            }
        }

        #endregion

        #region Helpers

        protected IContextData MergeContextData(IContextData localContextData)
        {
            if (localContextData == null)
                return GlobalContextData;

            if (GlobalContextData == null)
                return localContextData;

            IContextData merged = new ContextData();
            merged.ClaimValues = GlobalContextData.ClaimValues.Concat(localContextData.ClaimValues).ToList();
            return merged;
        }

        #endregion
    }
}

File: java/pca-client/src/main/java/com/sdl/web/pca/client/PublicContentApi.java
package com.sdl.web.pca.client;

import com.fasterxml.jackson.databind.JsonNode;
import com.sdl.web.pca.client.contentmodel.ContextData;
import com.sdl.web.pca.client.contentmodel.Pagination;
import com.sdl.web.pca.client.contentmodel.enums.ContentIncludeMode;
import com.sdl.web.pca.client.contentmodel.enums.ContentNamespace;
import com.sdl.web.pca.client.contentmodel.enums.ContentType;
import com.sdl.web.pca.client.contentmodel.enums.DataModelType;
import com.sdl.web.pca.client.contentmodel.enums.DcpType;
import com.sdl.web.pca.client.contentmodel.enums.PageInclusion;
import com.sdl.web.pca.client.contentmodel.generated.Ancestor;
import com.sdl.web.pca.client.contentmodel.generated.BinaryComponent;
import com.sdl.web.pca.client.contentmodel.generated.ComponentPresentation;
import com.sdl.web.pca.client.contentmodel.generated.ComponentPresentationConnection;
import com.sdl.web.pca.client.contentmodel.generated.InputComponentPresentationFilter;
import com.sdl.web.pca.client.contentmodel.generated.InputItemFilter;
import com.sdl.web.pca.client.contentmodel.generated.InputPublicationFilter;
import com.sdl.web.pca.client.contentmodel.generated.InputSortParam;
import com.sdl.web.pca.client.contentmodel.generated.ItemConnection;
import com.sdl.web.pca.client.contentmodel.generated.Page;
import com.sdl.web.pca.client.contentmodel.generated.PageConnection;
import com.sdl.web.pca.client.contentmodel.generated.Publication;
import com.sdl.web.pca.client.contentmodel.generated.PublicationConnection;
import com.sdl.web.pca.client.contentmodel.generated.PublicationMapping;
import com.sdl.web.pca.client.contentmodel.generated.TaxonomySitemapItem;
import com.sdl.web.pca.client.exception.PublicContentApiException;
import com.sdl.web.pca.client.util.CmUri;

/**
 * This interface enables java clients to connect to the GraphQL Service
 */
public interface PublicContentApi {

    /**
     * Retrieves ComponentPresentation object by given namespace, publication id, component id and template id.
     *
     * @param ns                 namespace
     * @param publicationId      publication id
     * @param componentId        component id
     * @param templateId         template id
     * @param customMetaFilter   custom meta filter
     * @param contentIncludeMode content include mode
     * @param contextData        context data
     * @return ComponentPresentation instance
     */
    ComponentPresentation getComponentPresentation(ContentNamespace ns, int publicationId, int componentId, int templateId,
                                                   String customMetaFilter, ContentIncludeMode contentIncludeMode,
                                                   ContextData contextData);

    /**
     * Retrieves ComponentPresentationConnection by given namespace, publication id and component presentation filter.
     *
     * @param ns                 namespace
     * @param publicationId      publication id
     * @param filter             component presentation filter
     * @param sort               sorting
     * @param pagination         pagination
     * @param customMetaFilter   custom meta filter
     * @param contentIncludeMode content include mode
     * @param contextData        context data
     * @return ComponentPresentationConnection instance
     */
    ComponentPresentationConnection getComponentPresentations(ContentNamespace ns, int publicationId,
                                                              InputComponentPresentationFilter filter, InputSortParam sort,
                                                              Pagination pagination, String customMetaFilter,
                                                              ContentIncludeMode contentIncludeMode, ContextData contextData);

    /**
     * Retrieves Page representation by given namespace, publication id and page id.
     *
     * @param ns                 namespace
     * @param publicationId      publication id
     * @param pageId             page id
     * @param customMetaFilter   custom meta filter
     * @param contentIncludeMode content include mode
     * @param contextData        context data
     * @return Page representation
     */
    Page getPage(ContentNamespace ns, int publicationId, int pageId, String customMetaFilter,
                 ContentIncludeMode contentIncludeMode, ContextData contextData);

    /**
     * Retrieves Page representation by given namespace, publication id and URL.
     *
     * @param ns                 namespace
     * @param publicationId      publication id
     * @param url                page URL
     * @param customMetaFilter   custom meta filter
     * @param contentIncludeMode content include mode
     * @param contextData        context data
     * @return Page representation
     */
    Page getPage(ContentNamespace ns, int publicationId, String url, String customMetaFilter,
                 ContentIncludeMode contentIncludeMode, ContextData contextData);

    /**
     * Retrieves Page representation by fiven CmUri.
     *
     * @param cmUri              CmUri of the page
     * @param customMetaFilter   custom meta filter
     * @param contentIncludeMode content include mode
     * @param contextData        context data
     * @return Page representation
     */
    Page getPage(CmUri cmUri, String customMetaFilter,
                 ContentIncludeMode contentIncludeMode, ContextData contextData);

    /**
     * Retrieves PageConnection representation by given namespace, pagination and URL
     *
     * @param ns                 namespace
     * @param pagination         pagination parameter
     * @param url                page URL
     * @param customMetaFilter   custom meta filter
     * @param contentIncludeMode content include mode
     * @param contextData        context data
     * @return PageConnection representation
     */
    PageConnection getPages(ContentNamespace ns, Pagination pagination, String url, String customMetaFilter,
                            ContentIncludeMode contentIncludeMode, ContextData contextData);

    /**
     * Retrieves BinaryComponent by providing publication id and binary id.
     *
     * @param ns            namespace
     * @param publicationId publication Id
     * @param binaryId      binary Id
     * @param contextData   context data
     * @return BinaryComponent representation
     * @throws PublicContentApiException
     */
    BinaryComponent getBinaryComponent(ContentNamespace ns, int publicationId, int binaryId, String customMetaFilter,
                                       ContextData contextData) throws PublicContentApiException;

    /**
     * Retrieves BinaryComponent by providing publication id and url of its binary component.
     *
     * @param ns            namespace
     * @param publicationId publication id
     * @param url           binary component url
     * @param contextData   context data
     * @return BinaryComponent representation
     * @throws PublicContentApiException
     */
    BinaryComponent getBinaryComponent(ContentNamespace ns, int publicationId, String url, String customMetaFilter,
                                       ContextData contextData) throws PublicContentApiException;

    /**
     * Retrieves BinaryConponent representation by providing its CMURI.
     *
     * @param cmUri       CMURI of binary component
     * @param contextData context data
     * @return BinaryComponent representation
     * @throws PublicContentApiException
     */
    BinaryComponent getBinaryComponent(CmUri cmUri, String customMetaFilter, ContextData contextData) throws PublicContentApiException;

    /**
     * Retrieves a data structure which holds list of {@code Item} implementation by providing filtering parameters.
     *
     * @param filter           specifies filtering parameters for item query
     * @param sort             defines sorting order
     * @param pagination       defines pagination parameter
     * @param contextData      context data
     * @param customMetaFilter custom metadata filter
     * @return ItemConnection object which holds Items
     * @throws PublicContentApiException
     */
    ItemConnection executeItemQuery(InputItemFilter filter, InputSortParam sort, Pagination pagination,
                                    String customMetaFilter, ContentIncludeMode contentIncludeMode,
                                    boolean includeContainerItems, ContextData contextData) throws PublicContentApiException;

    /**
     * Retrieves Publication by providing publication Id.
     *
     * @param ns               namespace
     * @param publicationId    publication id
     * @param contextData      context data
     * @param customMetaFilter custom metadata filter
     * @return Publication representation
     * @throws PublicContentApiException
     */
    Publication getPublication(ContentNamespace ns, int publicationId, String customMetaFilter,
                               ContextData contextData) throws PublicContentApiException;

    /**
     * Retrieves PublicationConnection object by given filter and pagination parameters.
     *
     * @param ns               namespace
     * @param pagination       defines pagination parameter
     * @param filter           defines filtering parameter
     * @param contextData      context data
     * @param customMetaFilter custom metadata filter
     * @return PublicationConnection representation
     */
    PublicationConnection getPublications(ContentNamespace ns, Pagination pagination, InputPublicationFilter filter,
                                          String customMetaFilter,
                                          ContextData contextData);

    /**
     * Retrieves link to page by given publication and page id.
     *
     * @param ns                 namespace
     * @param publicationId      publication id
     * @param pageId             page id
     * @param renderRelativeLink indicates if it is relative or full link
     * @return page link string
     * @throws PublicContentApiException
     */
    String resolvePageLink(ContentNamespace ns, int publicationId, int pageId, boolean renderRelativeLink) throws PublicContentApiException;

    /**
     * Retrieves link to component by given publication id and component id.
     *
     * @param ns                         namespace
     * @param publicationId              publication id
     * @param componentId                component id
     * @param sourcePageId               source page id
     * @param excludeComponentTemplateId exclude compoonent template id
     * @param renderRelativeLink         indicates if it is relative or full link
     * @return component link string
     * @throws PublicContentApiException
     */
    String resolveComponentLink(ContentNamespace ns, int publicationId, int componentId, Integer sourcePageId,
                                Integer excludeComponentTemplateId, boolean renderRelativeLink) throws PublicContentApiException;

    /**
     * Retrieves link to binary component by given publication id and binary id.
     *
     * @param ns                 namespace
     * @param publicationId      publication id
     * @param binaryId           binary id
     * @param variantId          variant id
     * @param renderRelativeLink indicates if it is relative or full link
     * @return binary link string
     * @throws PublicContentApiException
     */
    String resolveBinaryLink(ContentNamespace ns, int publicationId, int binaryId, String variantId,
                             boolean renderRelativeLink) throws PublicContentApiException;

    /**
     * Retrieves link to dynamic component by given publication, page, component and template ids.
     *
     * @param ns                 namespace
     * @param publicationId      publication id
     * @param pageId             page id
     * @param componentId        component id
     * @param templateId         template id
     * @param renderRelativeLink indicates if it is relative or full link
     * @return dynamic component link string
     * @throws PublicContentApiException
     */
    String resolveDynamicComponentLink(ContentNamespace ns, int publicationId, int pageId, int componentId,
                                       int templateId, boolean renderRelativeLink) throws PublicContentApiException;

    /**
     * Retrieves publication mapping by given url.
     *
     * @param ns  namespace
     * @param url url
     * @return PublicationMapping representation
     * @throws PublicContentApiException
     */
    PublicationMapping getPublicationMapping(ContentNamespace ns, String url) throws PublicContentApiException;

    /**
     * Retrieves page model data by given publication id and url.
     *
     * @param ns            namespace
     * @param publicationId publication id
     * @param url           url
     * @param contentType   content type
     * @param modelType     data model type
     * @param pageInclusion page inclusion
     * @param contextData   context data
     * @return json representation of page model
     * @throws PublicContentApiException
     */
    JsonNode getPageModelData(ContentNamespace ns, int publicationId, String url, ContentType contentType,
                              DataModelType modelType, PageInclusion pageInclusion, ContentIncludeMode contentIncludeMode,
                              ContextData contextData) throws PublicContentApiException;

    /**
     * Retrieves page model data by given publication id and page id.
     *
     * @param ns            namespace
     * @param publicationId publication id
     * @param pageId        page id
     * @param contentType   content type
     * @param modelType     data model type
     * @param pageInclusion page inclusion
     * @param contextData   context data
     * @return json representation of page model
     * @throws PublicContentApiException
     */
    JsonNode getPageModelData(ContentNamespace ns, int publicationId, int pageId, ContentType contentType,
                              DataModelType modelType, PageInclusion pageInclusion, ContentIncludeMode contentIncludeMode,
                              ContextData contextData) throws PublicContentApiException;

    /**
     * Retrieves entity model data by given publication, entity and template ids.
     *
     * @param ns            namespace
     * @param publicationId publication id
     * @param entityId      entity id
     * @param templateId    template id
     * @param contentType   content type
     * @param modelType     model type
     * @param dcpType       dcp type
     * @param contextData   context data
     * @return json representation of entity model
     * @throws PublicContentApiException
     */
    JsonNode getEntityModelData(ContentNamespace ns, int publicationId, int entityId, int templateId,
                                ContentType contentType,
                                DataModelType modelType, DcpType dcpType, ContentIncludeMode contentIncludeMode,
                                ContextData contextData) throws PublicContentApiException;

    /**
     * Retrieves TaxonomySitemapItem by given publication id.
     *
     * @param ns               namespace
     * @param publicationId    publication id
     * @param descendantLevels descendant level
     * @param contextData      context data
     * @return TaxonomySitemapItem representation
     * @throws PublicContentApiException
     */
    TaxonomySitemapItem getSitemap(ContentNamespace ns, int publicationId, int descendantLevels,
                                   ContextData contextData) throws PublicContentApiException;

    /**
     * Retrieves TaxonomySitemapItem array subtree by given publication id and parent taxonomy node id.
     *
     * @param ns               namespace
     * @param publicationId    publication id
     * @param taxonomyNodeId   taxonomy node id
     * @param descendantLevels descendant levels
     * @param contextData      context data
     * @return TaxonomySitemapItem array representation
     * @throws PublicContentApiException
     */
    TaxonomySitemapItem[] getSitemapSubtree(ContentNamespace ns, int publicationId, String taxonomyNodeId,
                                            int descendantLevels, Ancestor ancestor,
                                            ContextData contextData) throws PublicContentApiException;

}
